app/models/host.rb

```diff
@@ -1939,17 +1939,15 @@ def self.node_types
   end
 
   def self.openstack_hosts_exists?
-    ems = ManageIQ::Providers::Openstack::InfraManager.pluck(:id)
-    ems.empty? ? false : Host.where(:ems_id => ems).exists?
+    ManageIQ::Providers::Openstack::InfraManager::Host.count > 0
   end
 
   def self.non_openstack_hosts_exists?
-    ems = ManageIQ::Providers::Openstack::InfraManager.pluck(:id)
-    Host.where.not(:ems_id => ems).exists?
+    (Host.count - ManageIQ::Providers::Openstack::InfraManager::Host.count) > 0
   end
 
   def openstack_host?
-    ext_management_system.class == ManageIQ::Providers::Openstack::InfraManager
+    kind_of?(ManageIQ::Providers::Openstack::InfraManager::Host)
```

Comment:
@Fryguy : Yes, this is very non-pluggable. But it's driven by very specific product requirement where menu items are required to change name based on existence of certain types of hosts or clusters. The requirement was to do that right away when the condition changes.

To get this perform better we might cache these values and invalidate the them when Host or Cluster is added to the database. However I don't know how to do this if the particular records would be copied from another appliance

See `infrastructure_menu_section` in `app/presenters/menu/default_menu.rb`.